packages/zengine-sdk/src/filter-matcher.ts

```diff
@@ -0,0 +1,263 @@
+    }
+  },
+  ruleNotIn: function (recordValue: any, ruleValue: any) {
+    if (!Array.isArray(ruleValue)) {
+      return false
+    }
+    return !matchers.ruleIn(recordValue, ruleValue)
+  }
+}
+
+export const getRuleValues = function (rule: FieldRule) {
+  if (typeof rule.value === 'string' && rule.value.indexOf('|') !== -1) {
+    return rule.value.split('|')
+  }
+  if (rule.value === 'null' || rule.value === null) {
+    return ['']
+  }
+  return [rule.value]
+}
+
+export const getRecordValue = function (record: ZengineRecord, rule: FieldRule) {
+  const attributePieces: string[] = rule.attribute.split('.')
+  // Parse current record value of this rule's attribute, including dotted names (e.g. "folder.id")
+  let recordValue: any = record
+
+  attributePieces.forEach(function (attributePiece: string) {
+    recordValue = recordValue && recordValue[attributePiece]
+  })
+
+  // Parse subobject properties to use for check - e.g. field123.value for upload, field456.id for linked/member
+  if (recordValue) {
+    if (recordValue.value !== undefined) {
+      recordValue = recordValue.value
+    } else if (recordValue.id !== undefined) {
+      recordValue = recordValue.id
+    }
+  }
+
+  if (recordValue === null || recordValue === undefined) {
+    return ''
+  }
+
+  return recordValue
+}
+
+export const recordMatchesRule = function (record: ZengineRecord, rule: FieldRule, options: RuleOptions = {}) {
+  const operators = ['and', 'or']
+
+  if (operators.indexOf(Object.keys(rule)[0]) !== -1) {
+    // Rule contains "and"/"or" key - nested filter
+    return recordMatchesFilter(record, rule, options)
+  }
+
+  if (rule.filter !== undefined) {
+    if (options.subfiltering) {
+      const subRecord: any = record[rule.attribute]
+      return recordMatchesFilter(subRecord, rule.filter, options)
+    } else {
+      throw new Error('Subfilter matching is not supported')
+    }
+  }
+
+  if (typeof rule.value === 'string' && rule.value.split('|').indexOf('logged-in-user') !== -1) {
+    throw new Error('Dynamic filter conditions are not supported')
+  }
+
+  // From here, we know we have a normal rule with "prefix", "attribute", and "value" properties.
+  const recordValue = getRecordValue(record, rule)
+  const ruleValues = getRuleValues(rule)
+
+  // Run actual match logic based on rule prefix
+  const matchFunctionName: string = ruleFunctionMap[rule.prefix]
+  const matchFunction: Function = matchers[matchFunctionName]
+
+  for (let i in ruleValues) {
+    if (matchFunction(recordValue, ruleValues[i])) {
+      return true
+    }
+  }
+  // All ruleValues failed to match
+  return false
+}
+
+export const recordMatchesFilter = function (record: ZengineRecord, filter: FieldRule, options: RuleOptions = {}) {
+  const currentOperator: string = Object.keys(filter)[0]
+
+  // @ts-ignore
+  if (filter[currentOperator].length === 0) {
+    // Empty filter / no rules - considered a "match all"
+    return true
+  }
+
+  // @ts-ignore
```

Comment:
what is being ignored?

Reply:
` Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'FieldRule'.
  No index signature with a parameter of type 'string' was found on type 'FieldRule'.`